// @dbml/core/lib/parse/ANTLR/parsers/postgresql/PostgreSQLParserVisitor.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _antlr = _interopRequireDefault(require("antlr4"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); } // Generated from .\src\parse\ANTLR\postgresql\PostgreSQLParser.g4 by ANTLR 4.13.0
// jshint ignore: start
// This class defines a complete generic visitor for a parse tree produced by PostgreSQLParser.
var PostgreSQLParserVisitor = /*#__PURE__*/function (_antlr4$tree$ParseTre) {
  _inherits(PostgreSQLParserVisitor, _antlr4$tree$ParseTre);
  var _super = _createSuper(PostgreSQLParserVisitor);
  function PostgreSQLParserVisitor() {
    _classCallCheck(this, PostgreSQLParserVisitor);
    return _super.apply(this, arguments);
  }
  _createClass(PostgreSQLParserVisitor, [{
    key: "visitRoot",
    value:
    // Visit a parse tree produced by PostgreSQLParser#root.
    function visitRoot(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#plsqlroot.
  }, {
    key: "visitPlsqlroot",
    value: function visitPlsqlroot(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmtblock.
  }, {
    key: "visitStmtblock",
    value: function visitStmtblock(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmtmulti.
  }, {
    key: "visitStmtmulti",
    value: function visitStmtmulti(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt.
  }, {
    key: "visitStmt",
    value: function visitStmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#plsqlconsolecommand.
  }, {
    key: "visitPlsqlconsolecommand",
    value: function visitPlsqlconsolecommand(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#callstmt.
  }, {
    key: "visitCallstmt",
    value: function visitCallstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createrolestmt.
  }, {
    key: "visitCreaterolestmt",
    value: function visitCreaterolestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_with.
  }, {
    key: "visitOpt_with",
    value: function visitOpt_with(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optrolelist.
  }, {
    key: "visitOptrolelist",
    value: function visitOptrolelist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alteroptrolelist.
  }, {
    key: "visitAlteroptrolelist",
    value: function visitAlteroptrolelist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alteroptroleelem.
  }, {
    key: "visitAlteroptroleelem",
    value: function visitAlteroptroleelem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createoptroleelem.
  }, {
    key: "visitCreateoptroleelem",
    value: function visitCreateoptroleelem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createuserstmt.
  }, {
    key: "visitCreateuserstmt",
    value: function visitCreateuserstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterrolestmt.
  }, {
    key: "visitAlterrolestmt",
    value: function visitAlterrolestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_in_database.
  }, {
    key: "visitOpt_in_database",
    value: function visitOpt_in_database(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterrolesetstmt.
  }, {
    key: "visitAlterrolesetstmt",
    value: function visitAlterrolesetstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#droprolestmt.
  }, {
    key: "visitDroprolestmt",
    value: function visitDroprolestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#creategroupstmt.
  }, {
    key: "visitCreategroupstmt",
    value: function visitCreategroupstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altergroupstmt.
  }, {
    key: "visitAltergroupstmt",
    value: function visitAltergroupstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#add_drop.
  }, {
    key: "visitAdd_drop",
    value: function visitAdd_drop(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createschemastmt.
  }, {
    key: "visitCreateschemastmt",
    value: function visitCreateschemastmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optschemaname.
  }, {
    key: "visitOptschemaname",
    value: function visitOptschemaname(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optschemaeltlist.
  }, {
    key: "visitOptschemaeltlist",
    value: function visitOptschemaeltlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#schema_stmt.
  }, {
    key: "visitSchema_stmt",
    value: function visitSchema_stmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#variablesetstmt.
  }, {
    key: "visitVariablesetstmt",
    value: function visitVariablesetstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#set_rest.
  }, {
    key: "visitSet_rest",
    value: function visitSet_rest(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#generic_set.
  }, {
    key: "visitGeneric_set",
    value: function visitGeneric_set(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#set_rest_more.
  }, {
    key: "visitSet_rest_more",
    value: function visitSet_rest_more(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#var_name.
  }, {
    key: "visitVar_name",
    value: function visitVar_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#var_list.
  }, {
    key: "visitVar_list",
    value: function visitVar_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#var_value.
  }, {
    key: "visitVar_value",
    value: function visitVar_value(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#iso_level.
  }, {
    key: "visitIso_level",
    value: function visitIso_level(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_boolean_or_string.
  }, {
    key: "visitOpt_boolean_or_string",
    value: function visitOpt_boolean_or_string(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#zone_value.
  }, {
    key: "visitZone_value",
    value: function visitZone_value(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_encoding.
  }, {
    key: "visitOpt_encoding",
    value: function visitOpt_encoding(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#nonreservedword_or_sconst.
  }, {
    key: "visitNonreservedword_or_sconst",
    value: function visitNonreservedword_or_sconst(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#variableresetstmt.
  }, {
    key: "visitVariableresetstmt",
    value: function visitVariableresetstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reset_rest.
  }, {
    key: "visitReset_rest",
    value: function visitReset_rest(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#generic_reset.
  }, {
    key: "visitGeneric_reset",
    value: function visitGeneric_reset(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#setresetclause.
  }, {
    key: "visitSetresetclause",
    value: function visitSetresetclause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#functionsetresetclause.
  }, {
    key: "visitFunctionsetresetclause",
    value: function visitFunctionsetresetclause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#variableshowstmt.
  }, {
    key: "visitVariableshowstmt",
    value: function visitVariableshowstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constraintssetstmt.
  }, {
    key: "visitConstraintssetstmt",
    value: function visitConstraintssetstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constraints_set_list.
  }, {
    key: "visitConstraints_set_list",
    value: function visitConstraints_set_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constraints_set_mode.
  }, {
    key: "visitConstraints_set_mode",
    value: function visitConstraints_set_mode(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#checkpointstmt.
  }, {
    key: "visitCheckpointstmt",
    value: function visitCheckpointstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#discardstmt.
  }, {
    key: "visitDiscardstmt",
    value: function visitDiscardstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altertablestmt.
  }, {
    key: "visitAltertablestmt",
    value: function visitAltertablestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_table_cmds.
  }, {
    key: "visitAlter_table_cmds",
    value: function visitAlter_table_cmds(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#partition_cmd.
  }, {
    key: "visitPartition_cmd",
    value: function visitPartition_cmd(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#index_partition_cmd.
  }, {
    key: "visitIndex_partition_cmd",
    value: function visitIndex_partition_cmd(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_table_cmd.
  }, {
    key: "visitAlter_table_cmd",
    value: function visitAlter_table_cmd(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_column_default.
  }, {
    key: "visitAlter_column_default",
    value: function visitAlter_column_default(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_drop_behavior.
  }, {
    key: "visitOpt_drop_behavior",
    value: function visitOpt_drop_behavior(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_collate_clause.
  }, {
    key: "visitOpt_collate_clause",
    value: function visitOpt_collate_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_using.
  }, {
    key: "visitAlter_using",
    value: function visitAlter_using(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#replica_identity.
  }, {
    key: "visitReplica_identity",
    value: function visitReplica_identity(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reloptions.
  }, {
    key: "visitReloptions",
    value: function visitReloptions(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_reloptions.
  }, {
    key: "visitOpt_reloptions",
    value: function visitOpt_reloptions(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reloption_list.
  }, {
    key: "visitReloption_list",
    value: function visitReloption_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reloption_elem.
  }, {
    key: "visitReloption_elem",
    value: function visitReloption_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_identity_column_option_list.
  }, {
    key: "visitAlter_identity_column_option_list",
    value: function visitAlter_identity_column_option_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_identity_column_option.
  }, {
    key: "visitAlter_identity_column_option",
    value: function visitAlter_identity_column_option(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#partitionboundspec.
  }, {
    key: "visitPartitionboundspec",
    value: function visitPartitionboundspec(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#hash_partbound_elem.
  }, {
    key: "visitHash_partbound_elem",
    value: function visitHash_partbound_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#hash_partbound.
  }, {
    key: "visitHash_partbound",
    value: function visitHash_partbound(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altercompositetypestmt.
  }, {
    key: "visitAltercompositetypestmt",
    value: function visitAltercompositetypestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_type_cmds.
  }, {
    key: "visitAlter_type_cmds",
    value: function visitAlter_type_cmds(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_type_cmd.
  }, {
    key: "visitAlter_type_cmd",
    value: function visitAlter_type_cmd(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#closeportalstmt.
  }, {
    key: "visitCloseportalstmt",
    value: function visitCloseportalstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copystmt.
  }, {
    key: "visitCopystmt",
    value: function visitCopystmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_from.
  }, {
    key: "visitCopy_from",
    value: function visitCopy_from(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_program.
  }, {
    key: "visitOpt_program",
    value: function visitOpt_program(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_file_name.
  }, {
    key: "visitCopy_file_name",
    value: function visitCopy_file_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_options.
  }, {
    key: "visitCopy_options",
    value: function visitCopy_options(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_opt_list.
  }, {
    key: "visitCopy_opt_list",
    value: function visitCopy_opt_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_opt_item.
  }, {
    key: "visitCopy_opt_item",
    value: function visitCopy_opt_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_binary.
  }, {
    key: "visitOpt_binary",
    value: function visitOpt_binary(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_delimiter.
  }, {
    key: "visitCopy_delimiter",
    value: function visitCopy_delimiter(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_using.
  }, {
    key: "visitOpt_using",
    value: function visitOpt_using(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_generic_opt_list.
  }, {
    key: "visitCopy_generic_opt_list",
    value: function visitCopy_generic_opt_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_generic_opt_elem.
  }, {
    key: "visitCopy_generic_opt_elem",
    value: function visitCopy_generic_opt_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_generic_opt_arg.
  }, {
    key: "visitCopy_generic_opt_arg",
    value: function visitCopy_generic_opt_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_generic_opt_arg_list.
  }, {
    key: "visitCopy_generic_opt_arg_list",
    value: function visitCopy_generic_opt_arg_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#copy_generic_opt_arg_list_item.
  }, {
    key: "visitCopy_generic_opt_arg_list_item",
    value: function visitCopy_generic_opt_arg_list_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createstmt.
  }, {
    key: "visitCreatestmt",
    value: function visitCreatestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opttemp.
  }, {
    key: "visitOpttemp",
    value: function visitOpttemp(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opttableelementlist.
  }, {
    key: "visitOpttableelementlist",
    value: function visitOpttableelementlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opttypedtableelementlist.
  }, {
    key: "visitOpttypedtableelementlist",
    value: function visitOpttypedtableelementlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#tableelementlist.
  }, {
    key: "visitTableelementlist",
    value: function visitTableelementlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#typedtableelementlist.
  }, {
    key: "visitTypedtableelementlist",
    value: function visitTypedtableelementlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#tableelement.
  }, {
    key: "visitTableelement",
    value: function visitTableelement(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#typedtableelement.
  }, {
    key: "visitTypedtableelement",
    value: function visitTypedtableelement(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#columnDef.
  }, {
    key: "visitColumnDef",
    value: function visitColumnDef(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#columnOptions.
  }, {
    key: "visitColumnOptions",
    value: function visitColumnOptions(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#colquallist.
  }, {
    key: "visitColquallist",
    value: function visitColquallist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#colconstraint.
  }, {
    key: "visitColconstraint",
    value: function visitColconstraint(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#colconstraintelem.
  }, {
    key: "visitColconstraintelem",
    value: function visitColconstraintelem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#generated_when.
  }, {
    key: "visitGenerated_when",
    value: function visitGenerated_when(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constraintattr.
  }, {
    key: "visitConstraintattr",
    value: function visitConstraintattr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#tablelikeclause.
  }, {
    key: "visitTablelikeclause",
    value: function visitTablelikeclause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#tablelikeoptionlist.
  }, {
    key: "visitTablelikeoptionlist",
    value: function visitTablelikeoptionlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#tablelikeoption.
  }, {
    key: "visitTablelikeoption",
    value: function visitTablelikeoption(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#tableconstraint.
  }, {
    key: "visitTableconstraint",
    value: function visitTableconstraint(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constraintelem.
  }, {
    key: "visitConstraintelem",
    value: function visitConstraintelem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_no_inherit.
  }, {
    key: "visitOpt_no_inherit",
    value: function visitOpt_no_inherit(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_column_list.
  }, {
    key: "visitOpt_column_list",
    value: function visitOpt_column_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#columnlist.
  }, {
    key: "visitColumnlist",
    value: function visitColumnlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#columnElem.
  }, {
    key: "visitColumnElem",
    value: function visitColumnElem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_c_include.
  }, {
    key: "visitOpt_c_include",
    value: function visitOpt_c_include(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#key_match.
  }, {
    key: "visitKey_match",
    value: function visitKey_match(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#exclusionconstraintlist.
  }, {
    key: "visitExclusionconstraintlist",
    value: function visitExclusionconstraintlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#exclusionconstraintelem.
  }, {
    key: "visitExclusionconstraintelem",
    value: function visitExclusionconstraintelem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#exclusionwhereclause.
  }, {
    key: "visitExclusionwhereclause",
    value: function visitExclusionwhereclause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#key_actions.
  }, {
    key: "visitKey_actions",
    value: function visitKey_actions(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#key_update.
  }, {
    key: "visitKey_update",
    value: function visitKey_update(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#key_delete.
  }, {
    key: "visitKey_delete",
    value: function visitKey_delete(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#key_action.
  }, {
    key: "visitKey_action",
    value: function visitKey_action(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optinherit.
  }, {
    key: "visitOptinherit",
    value: function visitOptinherit(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optpartitionspec.
  }, {
    key: "visitOptpartitionspec",
    value: function visitOptpartitionspec(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#partitionspec.
  }, {
    key: "visitPartitionspec",
    value: function visitPartitionspec(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#part_params.
  }, {
    key: "visitPart_params",
    value: function visitPart_params(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#part_elem.
  }, {
    key: "visitPart_elem",
    value: function visitPart_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#table_access_method_clause.
  }, {
    key: "visitTable_access_method_clause",
    value: function visitTable_access_method_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optwith.
  }, {
    key: "visitOptwith",
    value: function visitOptwith(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#oncommitoption.
  }, {
    key: "visitOncommitoption",
    value: function visitOncommitoption(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opttablespace.
  }, {
    key: "visitOpttablespace",
    value: function visitOpttablespace(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optconstablespace.
  }, {
    key: "visitOptconstablespace",
    value: function visitOptconstablespace(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#existingindex.
  }, {
    key: "visitExistingindex",
    value: function visitExistingindex(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createstatsstmt.
  }, {
    key: "visitCreatestatsstmt",
    value: function visitCreatestatsstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterstatsstmt.
  }, {
    key: "visitAlterstatsstmt",
    value: function visitAlterstatsstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createasstmt.
  }, {
    key: "visitCreateasstmt",
    value: function visitCreateasstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#create_as_target.
  }, {
    key: "visitCreate_as_target",
    value: function visitCreate_as_target(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_with_data.
  }, {
    key: "visitOpt_with_data",
    value: function visitOpt_with_data(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#creatematviewstmt.
  }, {
    key: "visitCreatematviewstmt",
    value: function visitCreatematviewstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#create_mv_target.
  }, {
    key: "visitCreate_mv_target",
    value: function visitCreate_mv_target(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optnolog.
  }, {
    key: "visitOptnolog",
    value: function visitOptnolog(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#refreshmatviewstmt.
  }, {
    key: "visitRefreshmatviewstmt",
    value: function visitRefreshmatviewstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createseqstmt.
  }, {
    key: "visitCreateseqstmt",
    value: function visitCreateseqstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterseqstmt.
  }, {
    key: "visitAlterseqstmt",
    value: function visitAlterseqstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optseqoptlist.
  }, {
    key: "visitOptseqoptlist",
    value: function visitOptseqoptlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optparenthesizedseqoptlist.
  }, {
    key: "visitOptparenthesizedseqoptlist",
    value: function visitOptparenthesizedseqoptlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#seqoptlist.
  }, {
    key: "visitSeqoptlist",
    value: function visitSeqoptlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#seqoptelem.
  }, {
    key: "visitSeqoptelem",
    value: function visitSeqoptelem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_by.
  }, {
    key: "visitOpt_by",
    value: function visitOpt_by(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#numericonly.
  }, {
    key: "visitNumericonly",
    value: function visitNumericonly(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#numericonly_list.
  }, {
    key: "visitNumericonly_list",
    value: function visitNumericonly_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createplangstmt.
  }, {
    key: "visitCreateplangstmt",
    value: function visitCreateplangstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_trusted.
  }, {
    key: "visitOpt_trusted",
    value: function visitOpt_trusted(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#handler_name.
  }, {
    key: "visitHandler_name",
    value: function visitHandler_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_inline_handler.
  }, {
    key: "visitOpt_inline_handler",
    value: function visitOpt_inline_handler(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#validator_clause.
  }, {
    key: "visitValidator_clause",
    value: function visitValidator_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_validator.
  }, {
    key: "visitOpt_validator",
    value: function visitOpt_validator(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_procedural.
  }, {
    key: "visitOpt_procedural",
    value: function visitOpt_procedural(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createtablespacestmt.
  }, {
    key: "visitCreatetablespacestmt",
    value: function visitCreatetablespacestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opttablespaceowner.
  }, {
    key: "visitOpttablespaceowner",
    value: function visitOpttablespaceowner(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#droptablespacestmt.
  }, {
    key: "visitDroptablespacestmt",
    value: function visitDroptablespacestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createextensionstmt.
  }, {
    key: "visitCreateextensionstmt",
    value: function visitCreateextensionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#create_extension_opt_list.
  }, {
    key: "visitCreate_extension_opt_list",
    value: function visitCreate_extension_opt_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#create_extension_opt_item.
  }, {
    key: "visitCreate_extension_opt_item",
    value: function visitCreate_extension_opt_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterextensionstmt.
  }, {
    key: "visitAlterextensionstmt",
    value: function visitAlterextensionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_extension_opt_list.
  }, {
    key: "visitAlter_extension_opt_list",
    value: function visitAlter_extension_opt_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_extension_opt_item.
  }, {
    key: "visitAlter_extension_opt_item",
    value: function visitAlter_extension_opt_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterextensioncontentsstmt.
  }, {
    key: "visitAlterextensioncontentsstmt",
    value: function visitAlterextensioncontentsstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createfdwstmt.
  }, {
    key: "visitCreatefdwstmt",
    value: function visitCreatefdwstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#fdw_option.
  }, {
    key: "visitFdw_option",
    value: function visitFdw_option(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#fdw_options.
  }, {
    key: "visitFdw_options",
    value: function visitFdw_options(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_fdw_options.
  }, {
    key: "visitOpt_fdw_options",
    value: function visitOpt_fdw_options(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterfdwstmt.
  }, {
    key: "visitAlterfdwstmt",
    value: function visitAlterfdwstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#create_generic_options.
  }, {
    key: "visitCreate_generic_options",
    value: function visitCreate_generic_options(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#generic_option_list.
  }, {
    key: "visitGeneric_option_list",
    value: function visitGeneric_option_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_generic_options.
  }, {
    key: "visitAlter_generic_options",
    value: function visitAlter_generic_options(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_generic_option_list.
  }, {
    key: "visitAlter_generic_option_list",
    value: function visitAlter_generic_option_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alter_generic_option_elem.
  }, {
    key: "visitAlter_generic_option_elem",
    value: function visitAlter_generic_option_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#generic_option_elem.
  }, {
    key: "visitGeneric_option_elem",
    value: function visitGeneric_option_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#generic_option_name.
  }, {
    key: "visitGeneric_option_name",
    value: function visitGeneric_option_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#generic_option_arg.
  }, {
    key: "visitGeneric_option_arg",
    value: function visitGeneric_option_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createforeignserverstmt.
  }, {
    key: "visitCreateforeignserverstmt",
    value: function visitCreateforeignserverstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_type.
  }, {
    key: "visitOpt_type",
    value: function visitOpt_type(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#foreign_server_version.
  }, {
    key: "visitForeign_server_version",
    value: function visitForeign_server_version(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_foreign_server_version.
  }, {
    key: "visitOpt_foreign_server_version",
    value: function visitOpt_foreign_server_version(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterforeignserverstmt.
  }, {
    key: "visitAlterforeignserverstmt",
    value: function visitAlterforeignserverstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createforeigntablestmt.
  }, {
    key: "visitCreateforeigntablestmt",
    value: function visitCreateforeigntablestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#importforeignschemastmt.
  }, {
    key: "visitImportforeignschemastmt",
    value: function visitImportforeignschemastmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#import_qualification_type.
  }, {
    key: "visitImport_qualification_type",
    value: function visitImport_qualification_type(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#import_qualification.
  }, {
    key: "visitImport_qualification",
    value: function visitImport_qualification(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createusermappingstmt.
  }, {
    key: "visitCreateusermappingstmt",
    value: function visitCreateusermappingstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#auth_ident.
  }, {
    key: "visitAuth_ident",
    value: function visitAuth_ident(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dropusermappingstmt.
  }, {
    key: "visitDropusermappingstmt",
    value: function visitDropusermappingstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterusermappingstmt.
  }, {
    key: "visitAlterusermappingstmt",
    value: function visitAlterusermappingstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createpolicystmt.
  }, {
    key: "visitCreatepolicystmt",
    value: function visitCreatepolicystmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterpolicystmt.
  }, {
    key: "visitAlterpolicystmt",
    value: function visitAlterpolicystmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rowsecurityoptionalexpr.
  }, {
    key: "visitRowsecurityoptionalexpr",
    value: function visitRowsecurityoptionalexpr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rowsecurityoptionalwithcheck.
  }, {
    key: "visitRowsecurityoptionalwithcheck",
    value: function visitRowsecurityoptionalwithcheck(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rowsecuritydefaulttorole.
  }, {
    key: "visitRowsecuritydefaulttorole",
    value: function visitRowsecuritydefaulttorole(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rowsecurityoptionaltorole.
  }, {
    key: "visitRowsecurityoptionaltorole",
    value: function visitRowsecurityoptionaltorole(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rowsecuritydefaultpermissive.
  }, {
    key: "visitRowsecuritydefaultpermissive",
    value: function visitRowsecuritydefaultpermissive(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rowsecuritydefaultforcmd.
  }, {
    key: "visitRowsecuritydefaultforcmd",
    value: function visitRowsecuritydefaultforcmd(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#row_security_cmd.
  }, {
    key: "visitRow_security_cmd",
    value: function visitRow_security_cmd(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createamstmt.
  }, {
    key: "visitCreateamstmt",
    value: function visitCreateamstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#am_type.
  }, {
    key: "visitAm_type",
    value: function visitAm_type(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createtrigstmt.
  }, {
    key: "visitCreatetrigstmt",
    value: function visitCreatetrigstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggeractiontime.
  }, {
    key: "visitTriggeractiontime",
    value: function visitTriggeractiontime(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggerevents.
  }, {
    key: "visitTriggerevents",
    value: function visitTriggerevents(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggeroneevent.
  }, {
    key: "visitTriggeroneevent",
    value: function visitTriggeroneevent(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggerreferencing.
  }, {
    key: "visitTriggerreferencing",
    value: function visitTriggerreferencing(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggertransitions.
  }, {
    key: "visitTriggertransitions",
    value: function visitTriggertransitions(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggertransition.
  }, {
    key: "visitTriggertransition",
    value: function visitTriggertransition(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#transitionoldornew.
  }, {
    key: "visitTransitionoldornew",
    value: function visitTransitionoldornew(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#transitionrowortable.
  }, {
    key: "visitTransitionrowortable",
    value: function visitTransitionrowortable(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#transitionrelname.
  }, {
    key: "visitTransitionrelname",
    value: function visitTransitionrelname(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggerforspec.
  }, {
    key: "visitTriggerforspec",
    value: function visitTriggerforspec(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggerforopteach.
  }, {
    key: "visitTriggerforopteach",
    value: function visitTriggerforopteach(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggerfortype.
  }, {
    key: "visitTriggerfortype",
    value: function visitTriggerfortype(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggerwhen.
  }, {
    key: "visitTriggerwhen",
    value: function visitTriggerwhen(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#function_or_procedure.
  }, {
    key: "visitFunction_or_procedure",
    value: function visitFunction_or_procedure(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggerfuncargs.
  }, {
    key: "visitTriggerfuncargs",
    value: function visitTriggerfuncargs(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#triggerfuncarg.
  }, {
    key: "visitTriggerfuncarg",
    value: function visitTriggerfuncarg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#optconstrfromtable.
  }, {
    key: "visitOptconstrfromtable",
    value: function visitOptconstrfromtable(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constraintattributespec.
  }, {
    key: "visitConstraintattributespec",
    value: function visitConstraintattributespec(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constraintattributeElem.
  }, {
    key: "visitConstraintattributeElem",
    value: function visitConstraintattributeElem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createeventtrigstmt.
  }, {
    key: "visitCreateeventtrigstmt",
    value: function visitCreateeventtrigstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#event_trigger_when_list.
  }, {
    key: "visitEvent_trigger_when_list",
    value: function visitEvent_trigger_when_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#event_trigger_when_item.
  }, {
    key: "visitEvent_trigger_when_item",
    value: function visitEvent_trigger_when_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#event_trigger_value_list.
  }, {
    key: "visitEvent_trigger_value_list",
    value: function visitEvent_trigger_value_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altereventtrigstmt.
  }, {
    key: "visitAltereventtrigstmt",
    value: function visitAltereventtrigstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#enable_trigger.
  }, {
    key: "visitEnable_trigger",
    value: function visitEnable_trigger(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createassertionstmt.
  }, {
    key: "visitCreateassertionstmt",
    value: function visitCreateassertionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#definestmt.
  }, {
    key: "visitDefinestmt",
    value: function visitDefinestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#definition.
  }, {
    key: "visitDefinition",
    value: function visitDefinition(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#def_list.
  }, {
    key: "visitDef_list",
    value: function visitDef_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#def_elem.
  }, {
    key: "visitDef_elem",
    value: function visitDef_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#def_arg.
  }, {
    key: "visitDef_arg",
    value: function visitDef_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#old_aggr_definition.
  }, {
    key: "visitOld_aggr_definition",
    value: function visitOld_aggr_definition(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#old_aggr_list.
  }, {
    key: "visitOld_aggr_list",
    value: function visitOld_aggr_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#old_aggr_elem.
  }, {
    key: "visitOld_aggr_elem",
    value: function visitOld_aggr_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_enum_val_list.
  }, {
    key: "visitOpt_enum_val_list",
    value: function visitOpt_enum_val_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#enum_val_list.
  }, {
    key: "visitEnum_val_list",
    value: function visitEnum_val_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterenumstmt.
  }, {
    key: "visitAlterenumstmt",
    value: function visitAlterenumstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_if_not_exists.
  }, {
    key: "visitOpt_if_not_exists",
    value: function visitOpt_if_not_exists(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createopclassstmt.
  }, {
    key: "visitCreateopclassstmt",
    value: function visitCreateopclassstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opclass_item_list.
  }, {
    key: "visitOpclass_item_list",
    value: function visitOpclass_item_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opclass_item.
  }, {
    key: "visitOpclass_item",
    value: function visitOpclass_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_default.
  }, {
    key: "visitOpt_default",
    value: function visitOpt_default(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_opfamily.
  }, {
    key: "visitOpt_opfamily",
    value: function visitOpt_opfamily(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opclass_purpose.
  }, {
    key: "visitOpclass_purpose",
    value: function visitOpclass_purpose(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_recheck.
  }, {
    key: "visitOpt_recheck",
    value: function visitOpt_recheck(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createopfamilystmt.
  }, {
    key: "visitCreateopfamilystmt",
    value: function visitCreateopfamilystmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alteropfamilystmt.
  }, {
    key: "visitAlteropfamilystmt",
    value: function visitAlteropfamilystmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opclass_drop_list.
  }, {
    key: "visitOpclass_drop_list",
    value: function visitOpclass_drop_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opclass_drop.
  }, {
    key: "visitOpclass_drop",
    value: function visitOpclass_drop(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dropopclassstmt.
  }, {
    key: "visitDropopclassstmt",
    value: function visitDropopclassstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dropopfamilystmt.
  }, {
    key: "visitDropopfamilystmt",
    value: function visitDropopfamilystmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dropownedstmt.
  }, {
    key: "visitDropownedstmt",
    value: function visitDropownedstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reassignownedstmt.
  }, {
    key: "visitReassignownedstmt",
    value: function visitReassignownedstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dropstmt.
  }, {
    key: "visitDropstmt",
    value: function visitDropstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#object_type_any_name.
  }, {
    key: "visitObject_type_any_name",
    value: function visitObject_type_any_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#object_type_name.
  }, {
    key: "visitObject_type_name",
    value: function visitObject_type_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#drop_type_name.
  }, {
    key: "visitDrop_type_name",
    value: function visitDrop_type_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#object_type_name_on_any_name.
  }, {
    key: "visitObject_type_name_on_any_name",
    value: function visitObject_type_name_on_any_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#any_name_list.
  }, {
    key: "visitAny_name_list",
    value: function visitAny_name_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#any_name.
  }, {
    key: "visitAny_name",
    value: function visitAny_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#attrs.
  }, {
    key: "visitAttrs",
    value: function visitAttrs(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#type_name_list.
  }, {
    key: "visitType_name_list",
    value: function visitType_name_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#truncatestmt.
  }, {
    key: "visitTruncatestmt",
    value: function visitTruncatestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_restart_seqs.
  }, {
    key: "visitOpt_restart_seqs",
    value: function visitOpt_restart_seqs(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#commentstmt.
  }, {
    key: "visitCommentstmt",
    value: function visitCommentstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#comment_text.
  }, {
    key: "visitComment_text",
    value: function visitComment_text(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#seclabelstmt.
  }, {
    key: "visitSeclabelstmt",
    value: function visitSeclabelstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_provider.
  }, {
    key: "visitOpt_provider",
    value: function visitOpt_provider(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#security_label.
  }, {
    key: "visitSecurity_label",
    value: function visitSecurity_label(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#fetchstmt.
  }, {
    key: "visitFetchstmt",
    value: function visitFetchstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#fetch_args.
  }, {
    key: "visitFetch_args",
    value: function visitFetch_args(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#from_in.
  }, {
    key: "visitFrom_in",
    value: function visitFrom_in(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_from_in.
  }, {
    key: "visitOpt_from_in",
    value: function visitOpt_from_in(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#grantstmt.
  }, {
    key: "visitGrantstmt",
    value: function visitGrantstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#revokestmt.
  }, {
    key: "visitRevokestmt",
    value: function visitRevokestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#privileges.
  }, {
    key: "visitPrivileges",
    value: function visitPrivileges(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#privilege_list.
  }, {
    key: "visitPrivilege_list",
    value: function visitPrivilege_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#privilege.
  }, {
    key: "visitPrivilege",
    value: function visitPrivilege(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#privilege_target.
  }, {
    key: "visitPrivilege_target",
    value: function visitPrivilege_target(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#grantee_list.
  }, {
    key: "visitGrantee_list",
    value: function visitGrantee_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#grantee.
  }, {
    key: "visitGrantee",
    value: function visitGrantee(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_grant_grant_option.
  }, {
    key: "visitOpt_grant_grant_option",
    value: function visitOpt_grant_grant_option(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#grantrolestmt.
  }, {
    key: "visitGrantrolestmt",
    value: function visitGrantrolestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#revokerolestmt.
  }, {
    key: "visitRevokerolestmt",
    value: function visitRevokerolestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_grant_admin_option.
  }, {
    key: "visitOpt_grant_admin_option",
    value: function visitOpt_grant_admin_option(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_granted_by.
  }, {
    key: "visitOpt_granted_by",
    value: function visitOpt_granted_by(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterdefaultprivilegesstmt.
  }, {
    key: "visitAlterdefaultprivilegesstmt",
    value: function visitAlterdefaultprivilegesstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#defacloptionlist.
  }, {
    key: "visitDefacloptionlist",
    value: function visitDefacloptionlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#defacloption.
  }, {
    key: "visitDefacloption",
    value: function visitDefacloption(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#defaclaction.
  }, {
    key: "visitDefaclaction",
    value: function visitDefaclaction(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#defacl_privilege_target.
  }, {
    key: "visitDefacl_privilege_target",
    value: function visitDefacl_privilege_target(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#indexstmt.
  }, {
    key: "visitIndexstmt",
    value: function visitIndexstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_unique.
  }, {
    key: "visitOpt_unique",
    value: function visitOpt_unique(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_concurrently.
  }, {
    key: "visitOpt_concurrently",
    value: function visitOpt_concurrently(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_index_name.
  }, {
    key: "visitOpt_index_name",
    value: function visitOpt_index_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#access_method_clause.
  }, {
    key: "visitAccess_method_clause",
    value: function visitAccess_method_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#index_params.
  }, {
    key: "visitIndex_params",
    value: function visitIndex_params(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#index_elem_options.
  }, {
    key: "visitIndex_elem_options",
    value: function visitIndex_elem_options(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#index_elem.
  }, {
    key: "visitIndex_elem",
    value: function visitIndex_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_include.
  }, {
    key: "visitOpt_include",
    value: function visitOpt_include(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#index_including_params.
  }, {
    key: "visitIndex_including_params",
    value: function visitIndex_including_params(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_collate.
  }, {
    key: "visitOpt_collate",
    value: function visitOpt_collate(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_class.
  }, {
    key: "visitOpt_class",
    value: function visitOpt_class(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_asc_desc.
  }, {
    key: "visitOpt_asc_desc",
    value: function visitOpt_asc_desc(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_nulls_order.
  }, {
    key: "visitOpt_nulls_order",
    value: function visitOpt_nulls_order(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createfunctionstmt.
  }, {
    key: "visitCreatefunctionstmt",
    value: function visitCreatefunctionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_or_replace.
  }, {
    key: "visitOpt_or_replace",
    value: function visitOpt_or_replace(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_args.
  }, {
    key: "visitFunc_args",
    value: function visitFunc_args(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_args_list.
  }, {
    key: "visitFunc_args_list",
    value: function visitFunc_args_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#function_with_argtypes_list.
  }, {
    key: "visitFunction_with_argtypes_list",
    value: function visitFunction_with_argtypes_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#function_with_argtypes.
  }, {
    key: "visitFunction_with_argtypes",
    value: function visitFunction_with_argtypes(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_args_with_defaults.
  }, {
    key: "visitFunc_args_with_defaults",
    value: function visitFunc_args_with_defaults(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_args_with_defaults_list.
  }, {
    key: "visitFunc_args_with_defaults_list",
    value: function visitFunc_args_with_defaults_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_arg.
  }, {
    key: "visitFunc_arg",
    value: function visitFunc_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#arg_class.
  }, {
    key: "visitArg_class",
    value: function visitArg_class(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#param_name.
  }, {
    key: "visitParam_name",
    value: function visitParam_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_return.
  }, {
    key: "visitFunc_return",
    value: function visitFunc_return(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_type.
  }, {
    key: "visitFunc_type",
    value: function visitFunc_type(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_arg_with_default.
  }, {
    key: "visitFunc_arg_with_default",
    value: function visitFunc_arg_with_default(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#aggr_arg.
  }, {
    key: "visitAggr_arg",
    value: function visitAggr_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#aggr_args.
  }, {
    key: "visitAggr_args",
    value: function visitAggr_args(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#aggr_args_list.
  }, {
    key: "visitAggr_args_list",
    value: function visitAggr_args_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#aggregate_with_argtypes.
  }, {
    key: "visitAggregate_with_argtypes",
    value: function visitAggregate_with_argtypes(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#aggregate_with_argtypes_list.
  }, {
    key: "visitAggregate_with_argtypes_list",
    value: function visitAggregate_with_argtypes_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createfunc_opt_list.
  }, {
    key: "visitCreatefunc_opt_list",
    value: function visitCreatefunc_opt_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#common_func_opt_item.
  }, {
    key: "visitCommon_func_opt_item",
    value: function visitCommon_func_opt_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createfunc_opt_item.
  }, {
    key: "visitCreatefunc_opt_item",
    value: function visitCreatefunc_opt_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_as.
  }, {
    key: "visitFunc_as",
    value: function visitFunc_as(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#transform_type_list.
  }, {
    key: "visitTransform_type_list",
    value: function visitTransform_type_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_definition.
  }, {
    key: "visitOpt_definition",
    value: function visitOpt_definition(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#table_func_column.
  }, {
    key: "visitTable_func_column",
    value: function visitTable_func_column(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#table_func_column_list.
  }, {
    key: "visitTable_func_column_list",
    value: function visitTable_func_column_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterfunctionstmt.
  }, {
    key: "visitAlterfunctionstmt",
    value: function visitAlterfunctionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterfunc_opt_list.
  }, {
    key: "visitAlterfunc_opt_list",
    value: function visitAlterfunc_opt_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_restrict.
  }, {
    key: "visitOpt_restrict",
    value: function visitOpt_restrict(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#removefuncstmt.
  }, {
    key: "visitRemovefuncstmt",
    value: function visitRemovefuncstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#removeaggrstmt.
  }, {
    key: "visitRemoveaggrstmt",
    value: function visitRemoveaggrstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#removeoperstmt.
  }, {
    key: "visitRemoveoperstmt",
    value: function visitRemoveoperstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#oper_argtypes.
  }, {
    key: "visitOper_argtypes",
    value: function visitOper_argtypes(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#any_operator.
  }, {
    key: "visitAny_operator",
    value: function visitAny_operator(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#operator_with_argtypes_list.
  }, {
    key: "visitOperator_with_argtypes_list",
    value: function visitOperator_with_argtypes_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#operator_with_argtypes.
  }, {
    key: "visitOperator_with_argtypes",
    value: function visitOperator_with_argtypes(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dostmt.
  }, {
    key: "visitDostmt",
    value: function visitDostmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dostmt_opt_list.
  }, {
    key: "visitDostmt_opt_list",
    value: function visitDostmt_opt_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dostmt_opt_item.
  }, {
    key: "visitDostmt_opt_item",
    value: function visitDostmt_opt_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createcaststmt.
  }, {
    key: "visitCreatecaststmt",
    value: function visitCreatecaststmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#cast_context.
  }, {
    key: "visitCast_context",
    value: function visitCast_context(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dropcaststmt.
  }, {
    key: "visitDropcaststmt",
    value: function visitDropcaststmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_if_exists.
  }, {
    key: "visitOpt_if_exists",
    value: function visitOpt_if_exists(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createtransformstmt.
  }, {
    key: "visitCreatetransformstmt",
    value: function visitCreatetransformstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#transform_element_list.
  }, {
    key: "visitTransform_element_list",
    value: function visitTransform_element_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#droptransformstmt.
  }, {
    key: "visitDroptransformstmt",
    value: function visitDroptransformstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reindexstmt.
  }, {
    key: "visitReindexstmt",
    value: function visitReindexstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reindex_target_type.
  }, {
    key: "visitReindex_target_type",
    value: function visitReindex_target_type(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reindex_target_multitable.
  }, {
    key: "visitReindex_target_multitable",
    value: function visitReindex_target_multitable(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reindex_option_list.
  }, {
    key: "visitReindex_option_list",
    value: function visitReindex_option_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reindex_option_elem.
  }, {
    key: "visitReindex_option_elem",
    value: function visitReindex_option_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altertblspcstmt.
  }, {
    key: "visitAltertblspcstmt",
    value: function visitAltertblspcstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#renamestmt.
  }, {
    key: "visitRenamestmt",
    value: function visitRenamestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_column.
  }, {
    key: "visitOpt_column",
    value: function visitOpt_column(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_set_data.
  }, {
    key: "visitOpt_set_data",
    value: function visitOpt_set_data(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterobjectdependsstmt.
  }, {
    key: "visitAlterobjectdependsstmt",
    value: function visitAlterobjectdependsstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_no.
  }, {
    key: "visitOpt_no",
    value: function visitOpt_no(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterobjectschemastmt.
  }, {
    key: "visitAlterobjectschemastmt",
    value: function visitAlterobjectschemastmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alteroperatorstmt.
  }, {
    key: "visitAlteroperatorstmt",
    value: function visitAlteroperatorstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#operator_def_list.
  }, {
    key: "visitOperator_def_list",
    value: function visitOperator_def_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#operator_def_elem.
  }, {
    key: "visitOperator_def_elem",
    value: function visitOperator_def_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#operator_def_arg.
  }, {
    key: "visitOperator_def_arg",
    value: function visitOperator_def_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altertypestmt.
  }, {
    key: "visitAltertypestmt",
    value: function visitAltertypestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterownerstmt.
  }, {
    key: "visitAlterownerstmt",
    value: function visitAlterownerstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createpublicationstmt.
  }, {
    key: "visitCreatepublicationstmt",
    value: function visitCreatepublicationstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_publication_for_tables.
  }, {
    key: "visitOpt_publication_for_tables",
    value: function visitOpt_publication_for_tables(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#publication_for_tables.
  }, {
    key: "visitPublication_for_tables",
    value: function visitPublication_for_tables(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterpublicationstmt.
  }, {
    key: "visitAlterpublicationstmt",
    value: function visitAlterpublicationstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createsubscriptionstmt.
  }, {
    key: "visitCreatesubscriptionstmt",
    value: function visitCreatesubscriptionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#publication_name_list.
  }, {
    key: "visitPublication_name_list",
    value: function visitPublication_name_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#publication_name_item.
  }, {
    key: "visitPublication_name_item",
    value: function visitPublication_name_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altersubscriptionstmt.
  }, {
    key: "visitAltersubscriptionstmt",
    value: function visitAltersubscriptionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dropsubscriptionstmt.
  }, {
    key: "visitDropsubscriptionstmt",
    value: function visitDropsubscriptionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rulestmt.
  }, {
    key: "visitRulestmt",
    value: function visitRulestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#ruleactionlist.
  }, {
    key: "visitRuleactionlist",
    value: function visitRuleactionlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#ruleactionmulti.
  }, {
    key: "visitRuleactionmulti",
    value: function visitRuleactionmulti(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#ruleactionstmt.
  }, {
    key: "visitRuleactionstmt",
    value: function visitRuleactionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#ruleactionstmtOrEmpty.
  }, {
    key: "visitRuleactionstmtOrEmpty",
    value: function visitRuleactionstmtOrEmpty(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#event.
  }, {
    key: "visitEvent",
    value: function visitEvent(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_instead.
  }, {
    key: "visitOpt_instead",
    value: function visitOpt_instead(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#notifystmt.
  }, {
    key: "visitNotifystmt",
    value: function visitNotifystmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#notify_payload.
  }, {
    key: "visitNotify_payload",
    value: function visitNotify_payload(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#listenstmt.
  }, {
    key: "visitListenstmt",
    value: function visitListenstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#unlistenstmt.
  }, {
    key: "visitUnlistenstmt",
    value: function visitUnlistenstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#transactionstmt.
  }, {
    key: "visitTransactionstmt",
    value: function visitTransactionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_transaction.
  }, {
    key: "visitOpt_transaction",
    value: function visitOpt_transaction(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#transaction_mode_item.
  }, {
    key: "visitTransaction_mode_item",
    value: function visitTransaction_mode_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#transaction_mode_list.
  }, {
    key: "visitTransaction_mode_list",
    value: function visitTransaction_mode_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#transaction_mode_list_or_empty.
  }, {
    key: "visitTransaction_mode_list_or_empty",
    value: function visitTransaction_mode_list_or_empty(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_transaction_chain.
  }, {
    key: "visitOpt_transaction_chain",
    value: function visitOpt_transaction_chain(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#viewstmt.
  }, {
    key: "visitViewstmt",
    value: function visitViewstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_check_option.
  }, {
    key: "visitOpt_check_option",
    value: function visitOpt_check_option(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#loadstmt.
  }, {
    key: "visitLoadstmt",
    value: function visitLoadstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createdbstmt.
  }, {
    key: "visitCreatedbstmt",
    value: function visitCreatedbstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createdb_opt_list.
  }, {
    key: "visitCreatedb_opt_list",
    value: function visitCreatedb_opt_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createdb_opt_items.
  }, {
    key: "visitCreatedb_opt_items",
    value: function visitCreatedb_opt_items(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createdb_opt_item.
  }, {
    key: "visitCreatedb_opt_item",
    value: function visitCreatedb_opt_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createdb_opt_name.
  }, {
    key: "visitCreatedb_opt_name",
    value: function visitCreatedb_opt_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_equal.
  }, {
    key: "visitOpt_equal",
    value: function visitOpt_equal(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterdatabasestmt.
  }, {
    key: "visitAlterdatabasestmt",
    value: function visitAlterdatabasestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterdatabasesetstmt.
  }, {
    key: "visitAlterdatabasesetstmt",
    value: function visitAlterdatabasesetstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#dropdbstmt.
  }, {
    key: "visitDropdbstmt",
    value: function visitDropdbstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#drop_option_list.
  }, {
    key: "visitDrop_option_list",
    value: function visitDrop_option_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#drop_option.
  }, {
    key: "visitDrop_option",
    value: function visitDrop_option(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altercollationstmt.
  }, {
    key: "visitAltercollationstmt",
    value: function visitAltercollationstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altersystemstmt.
  }, {
    key: "visitAltersystemstmt",
    value: function visitAltersystemstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createdomainstmt.
  }, {
    key: "visitCreatedomainstmt",
    value: function visitCreatedomainstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alterdomainstmt.
  }, {
    key: "visitAlterdomainstmt",
    value: function visitAlterdomainstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_as.
  }, {
    key: "visitOpt_as",
    value: function visitOpt_as(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altertsdictionarystmt.
  }, {
    key: "visitAltertsdictionarystmt",
    value: function visitAltertsdictionarystmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#altertsconfigurationstmt.
  }, {
    key: "visitAltertsconfigurationstmt",
    value: function visitAltertsconfigurationstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#any_with.
  }, {
    key: "visitAny_with",
    value: function visitAny_with(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#createconversionstmt.
  }, {
    key: "visitCreateconversionstmt",
    value: function visitCreateconversionstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#clusterstmt.
  }, {
    key: "visitClusterstmt",
    value: function visitClusterstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#cluster_index_specification.
  }, {
    key: "visitCluster_index_specification",
    value: function visitCluster_index_specification(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#vacuumstmt.
  }, {
    key: "visitVacuumstmt",
    value: function visitVacuumstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#analyzestmt.
  }, {
    key: "visitAnalyzestmt",
    value: function visitAnalyzestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#vac_analyze_option_list.
  }, {
    key: "visitVac_analyze_option_list",
    value: function visitVac_analyze_option_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#analyze_keyword.
  }, {
    key: "visitAnalyze_keyword",
    value: function visitAnalyze_keyword(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#vac_analyze_option_elem.
  }, {
    key: "visitVac_analyze_option_elem",
    value: function visitVac_analyze_option_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#vac_analyze_option_name.
  }, {
    key: "visitVac_analyze_option_name",
    value: function visitVac_analyze_option_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#vac_analyze_option_arg.
  }, {
    key: "visitVac_analyze_option_arg",
    value: function visitVac_analyze_option_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_analyze.
  }, {
    key: "visitOpt_analyze",
    value: function visitOpt_analyze(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_verbose.
  }, {
    key: "visitOpt_verbose",
    value: function visitOpt_verbose(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_full.
  }, {
    key: "visitOpt_full",
    value: function visitOpt_full(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_freeze.
  }, {
    key: "visitOpt_freeze",
    value: function visitOpt_freeze(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_name_list.
  }, {
    key: "visitOpt_name_list",
    value: function visitOpt_name_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#vacuum_relation.
  }, {
    key: "visitVacuum_relation",
    value: function visitVacuum_relation(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#vacuum_relation_list.
  }, {
    key: "visitVacuum_relation_list",
    value: function visitVacuum_relation_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_vacuum_relation_list.
  }, {
    key: "visitOpt_vacuum_relation_list",
    value: function visitOpt_vacuum_relation_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#explainstmt.
  }, {
    key: "visitExplainstmt",
    value: function visitExplainstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#explainablestmt.
  }, {
    key: "visitExplainablestmt",
    value: function visitExplainablestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#explain_option_list.
  }, {
    key: "visitExplain_option_list",
    value: function visitExplain_option_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#explain_option_elem.
  }, {
    key: "visitExplain_option_elem",
    value: function visitExplain_option_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#explain_option_name.
  }, {
    key: "visitExplain_option_name",
    value: function visitExplain_option_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#explain_option_arg.
  }, {
    key: "visitExplain_option_arg",
    value: function visitExplain_option_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#preparestmt.
  }, {
    key: "visitPreparestmt",
    value: function visitPreparestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#prep_type_clause.
  }, {
    key: "visitPrep_type_clause",
    value: function visitPrep_type_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#preparablestmt.
  }, {
    key: "visitPreparablestmt",
    value: function visitPreparablestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#executestmt.
  }, {
    key: "visitExecutestmt",
    value: function visitExecutestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#execute_param_clause.
  }, {
    key: "visitExecute_param_clause",
    value: function visitExecute_param_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#deallocatestmt.
  }, {
    key: "visitDeallocatestmt",
    value: function visitDeallocatestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#insertstmt.
  }, {
    key: "visitInsertstmt",
    value: function visitInsertstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#insert_target.
  }, {
    key: "visitInsert_target",
    value: function visitInsert_target(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#insert_rest.
  }, {
    key: "visitInsert_rest",
    value: function visitInsert_rest(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#override_kind.
  }, {
    key: "visitOverride_kind",
    value: function visitOverride_kind(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#insert_column_list.
  }, {
    key: "visitInsert_column_list",
    value: function visitInsert_column_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#insert_column_item.
  }, {
    key: "visitInsert_column_item",
    value: function visitInsert_column_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_on_conflict.
  }, {
    key: "visitOpt_on_conflict",
    value: function visitOpt_on_conflict(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_conf_expr.
  }, {
    key: "visitOpt_conf_expr",
    value: function visitOpt_conf_expr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#returning_clause.
  }, {
    key: "visitReturning_clause",
    value: function visitReturning_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#mergestmt.
  }, {
    key: "visitMergestmt",
    value: function visitMergestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#merge_insert_clause.
  }, {
    key: "visitMerge_insert_clause",
    value: function visitMerge_insert_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#merge_update_clause.
  }, {
    key: "visitMerge_update_clause",
    value: function visitMerge_update_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#merge_delete_clause.
  }, {
    key: "visitMerge_delete_clause",
    value: function visitMerge_delete_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#deletestmt.
  }, {
    key: "visitDeletestmt",
    value: function visitDeletestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#using_clause.
  }, {
    key: "visitUsing_clause",
    value: function visitUsing_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#lockstmt.
  }, {
    key: "visitLockstmt",
    value: function visitLockstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_lock.
  }, {
    key: "visitOpt_lock",
    value: function visitOpt_lock(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#lock_type.
  }, {
    key: "visitLock_type",
    value: function visitLock_type(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_nowait.
  }, {
    key: "visitOpt_nowait",
    value: function visitOpt_nowait(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_nowait_or_skip.
  }, {
    key: "visitOpt_nowait_or_skip",
    value: function visitOpt_nowait_or_skip(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#updatestmt.
  }, {
    key: "visitUpdatestmt",
    value: function visitUpdatestmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#set_clause_list.
  }, {
    key: "visitSet_clause_list",
    value: function visitSet_clause_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#set_clause.
  }, {
    key: "visitSet_clause",
    value: function visitSet_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#set_target.
  }, {
    key: "visitSet_target",
    value: function visitSet_target(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#set_target_list.
  }, {
    key: "visitSet_target_list",
    value: function visitSet_target_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#declarecursorstmt.
  }, {
    key: "visitDeclarecursorstmt",
    value: function visitDeclarecursorstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#cursor_name.
  }, {
    key: "visitCursor_name",
    value: function visitCursor_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#cursor_options.
  }, {
    key: "visitCursor_options",
    value: function visitCursor_options(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_hold.
  }, {
    key: "visitOpt_hold",
    value: function visitOpt_hold(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#selectstmt.
  }, {
    key: "visitSelectstmt",
    value: function visitSelectstmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#select_with_parens.
  }, {
    key: "visitSelect_with_parens",
    value: function visitSelect_with_parens(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#select_no_parens.
  }, {
    key: "visitSelect_no_parens",
    value: function visitSelect_no_parens(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#select_clause.
  }, {
    key: "visitSelect_clause",
    value: function visitSelect_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#simple_select.
  }, {
    key: "visitSimple_select",
    value: function visitSimple_select(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#union.
  }, {
    key: "visitUnion",
    value: function visitUnion(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#intersect.
  }, {
    key: "visitIntersect",
    value: function visitIntersect(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#except.
  }, {
    key: "visitExcept",
    value: function visitExcept(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#set_operator_with_all_or_distinct.
  }, {
    key: "visitSet_operator_with_all_or_distinct",
    value: function visitSet_operator_with_all_or_distinct(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#with_clause.
  }, {
    key: "visitWith_clause",
    value: function visitWith_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#cte_list.
  }, {
    key: "visitCte_list",
    value: function visitCte_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#common_table_expr.
  }, {
    key: "visitCommon_table_expr",
    value: function visitCommon_table_expr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_materialized.
  }, {
    key: "visitOpt_materialized",
    value: function visitOpt_materialized(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_with_clause.
  }, {
    key: "visitOpt_with_clause",
    value: function visitOpt_with_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#into_clause.
  }, {
    key: "visitInto_clause",
    value: function visitInto_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_strict.
  }, {
    key: "visitOpt_strict",
    value: function visitOpt_strict(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opttempTableName.
  }, {
    key: "visitOpttempTableName",
    value: function visitOpttempTableName(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_table.
  }, {
    key: "visitOpt_table",
    value: function visitOpt_table(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#all_or_distinct.
  }, {
    key: "visitAll_or_distinct",
    value: function visitAll_or_distinct(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#distinct_clause.
  }, {
    key: "visitDistinct_clause",
    value: function visitDistinct_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_all_clause.
  }, {
    key: "visitOpt_all_clause",
    value: function visitOpt_all_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_sort_clause.
  }, {
    key: "visitOpt_sort_clause",
    value: function visitOpt_sort_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#sort_clause.
  }, {
    key: "visitSort_clause",
    value: function visitSort_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#sortby_list.
  }, {
    key: "visitSortby_list",
    value: function visitSortby_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#sortby.
  }, {
    key: "visitSortby",
    value: function visitSortby(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#select_limit.
  }, {
    key: "visitSelect_limit",
    value: function visitSelect_limit(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_select_limit.
  }, {
    key: "visitOpt_select_limit",
    value: function visitOpt_select_limit(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#limit_clause.
  }, {
    key: "visitLimit_clause",
    value: function visitLimit_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#offset_clause.
  }, {
    key: "visitOffset_clause",
    value: function visitOffset_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#select_limit_value.
  }, {
    key: "visitSelect_limit_value",
    value: function visitSelect_limit_value(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#select_offset_value.
  }, {
    key: "visitSelect_offset_value",
    value: function visitSelect_offset_value(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#select_fetch_first_value.
  }, {
    key: "visitSelect_fetch_first_value",
    value: function visitSelect_fetch_first_value(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#i_or_f_const.
  }, {
    key: "visitI_or_f_const",
    value: function visitI_or_f_const(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#row_or_rows.
  }, {
    key: "visitRow_or_rows",
    value: function visitRow_or_rows(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#first_or_next.
  }, {
    key: "visitFirst_or_next",
    value: function visitFirst_or_next(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#group_clause.
  }, {
    key: "visitGroup_clause",
    value: function visitGroup_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#group_by_list.
  }, {
    key: "visitGroup_by_list",
    value: function visitGroup_by_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#group_by_item.
  }, {
    key: "visitGroup_by_item",
    value: function visitGroup_by_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#empty_grouping_set.
  }, {
    key: "visitEmpty_grouping_set",
    value: function visitEmpty_grouping_set(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rollup_clause.
  }, {
    key: "visitRollup_clause",
    value: function visitRollup_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#cube_clause.
  }, {
    key: "visitCube_clause",
    value: function visitCube_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#grouping_sets_clause.
  }, {
    key: "visitGrouping_sets_clause",
    value: function visitGrouping_sets_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#having_clause.
  }, {
    key: "visitHaving_clause",
    value: function visitHaving_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#for_locking_clause.
  }, {
    key: "visitFor_locking_clause",
    value: function visitFor_locking_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_for_locking_clause.
  }, {
    key: "visitOpt_for_locking_clause",
    value: function visitOpt_for_locking_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#for_locking_items.
  }, {
    key: "visitFor_locking_items",
    value: function visitFor_locking_items(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#for_locking_item.
  }, {
    key: "visitFor_locking_item",
    value: function visitFor_locking_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#for_locking_strength.
  }, {
    key: "visitFor_locking_strength",
    value: function visitFor_locking_strength(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#locked_rels_list.
  }, {
    key: "visitLocked_rels_list",
    value: function visitLocked_rels_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#values_clause.
  }, {
    key: "visitValues_clause",
    value: function visitValues_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#from_clause.
  }, {
    key: "visitFrom_clause",
    value: function visitFrom_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#from_list.
  }, {
    key: "visitFrom_list",
    value: function visitFrom_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#non_ansi_join.
  }, {
    key: "visitNon_ansi_join",
    value: function visitNon_ansi_join(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#table_ref.
  }, {
    key: "visitTable_ref",
    value: function visitTable_ref(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#alias_clause.
  }, {
    key: "visitAlias_clause",
    value: function visitAlias_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_alias_clause.
  }, {
    key: "visitOpt_alias_clause",
    value: function visitOpt_alias_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#table_alias_clause.
  }, {
    key: "visitTable_alias_clause",
    value: function visitTable_alias_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_alias_clause.
  }, {
    key: "visitFunc_alias_clause",
    value: function visitFunc_alias_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#join_type.
  }, {
    key: "visitJoin_type",
    value: function visitJoin_type(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#join_qual.
  }, {
    key: "visitJoin_qual",
    value: function visitJoin_qual(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#relation_expr.
  }, {
    key: "visitRelation_expr",
    value: function visitRelation_expr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#relation_expr_list.
  }, {
    key: "visitRelation_expr_list",
    value: function visitRelation_expr_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#relation_expr_opt_alias.
  }, {
    key: "visitRelation_expr_opt_alias",
    value: function visitRelation_expr_opt_alias(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#tablesample_clause.
  }, {
    key: "visitTablesample_clause",
    value: function visitTablesample_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_repeatable_clause.
  }, {
    key: "visitOpt_repeatable_clause",
    value: function visitOpt_repeatable_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_table.
  }, {
    key: "visitFunc_table",
    value: function visitFunc_table(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rowsfrom_item.
  }, {
    key: "visitRowsfrom_item",
    value: function visitRowsfrom_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rowsfrom_list.
  }, {
    key: "visitRowsfrom_list",
    value: function visitRowsfrom_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_col_def_list.
  }, {
    key: "visitOpt_col_def_list",
    value: function visitOpt_col_def_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_ordinality.
  }, {
    key: "visitOpt_ordinality",
    value: function visitOpt_ordinality(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#where_clause.
  }, {
    key: "visitWhere_clause",
    value: function visitWhere_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#where_or_current_clause.
  }, {
    key: "visitWhere_or_current_clause",
    value: function visitWhere_or_current_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opttablefuncelementlist.
  }, {
    key: "visitOpttablefuncelementlist",
    value: function visitOpttablefuncelementlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#tablefuncelementlist.
  }, {
    key: "visitTablefuncelementlist",
    value: function visitTablefuncelementlist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#tablefuncelement.
  }, {
    key: "visitTablefuncelement",
    value: function visitTablefuncelement(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xmltable.
  }, {
    key: "visitXmltable",
    value: function visitXmltable(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xmltable_column_list.
  }, {
    key: "visitXmltable_column_list",
    value: function visitXmltable_column_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xmltable_column_el.
  }, {
    key: "visitXmltable_column_el",
    value: function visitXmltable_column_el(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xmltable_column_option_list.
  }, {
    key: "visitXmltable_column_option_list",
    value: function visitXmltable_column_option_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xmltable_column_option_el.
  }, {
    key: "visitXmltable_column_option_el",
    value: function visitXmltable_column_option_el(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xml_namespace_list.
  }, {
    key: "visitXml_namespace_list",
    value: function visitXml_namespace_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xml_namespace_el.
  }, {
    key: "visitXml_namespace_el",
    value: function visitXml_namespace_el(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#typename.
  }, {
    key: "visitTypename",
    value: function visitTypename(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_array_bounds.
  }, {
    key: "visitOpt_array_bounds",
    value: function visitOpt_array_bounds(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#simpletypename.
  }, {
    key: "visitSimpletypename",
    value: function visitSimpletypename(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#consttypename.
  }, {
    key: "visitConsttypename",
    value: function visitConsttypename(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#generictype.
  }, {
    key: "visitGenerictype",
    value: function visitGenerictype(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_type_modifiers.
  }, {
    key: "visitOpt_type_modifiers",
    value: function visitOpt_type_modifiers(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#numeric.
  }, {
    key: "visitNumeric",
    value: function visitNumeric(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_float.
  }, {
    key: "visitOpt_float",
    value: function visitOpt_float(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#bit.
  }, {
    key: "visitBit",
    value: function visitBit(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constbit.
  }, {
    key: "visitConstbit",
    value: function visitConstbit(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#bitwithlength.
  }, {
    key: "visitBitwithlength",
    value: function visitBitwithlength(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#bitwithoutlength.
  }, {
    key: "visitBitwithoutlength",
    value: function visitBitwithoutlength(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#character.
  }, {
    key: "visitCharacter",
    value: function visitCharacter(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constcharacter.
  }, {
    key: "visitConstcharacter",
    value: function visitConstcharacter(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#character_c.
  }, {
    key: "visitCharacter_c",
    value: function visitCharacter_c(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_varying.
  }, {
    key: "visitOpt_varying",
    value: function visitOpt_varying(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constdatetime.
  }, {
    key: "visitConstdatetime",
    value: function visitConstdatetime(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#constinterval.
  }, {
    key: "visitConstinterval",
    value: function visitConstinterval(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_timezone.
  }, {
    key: "visitOpt_timezone",
    value: function visitOpt_timezone(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_interval.
  }, {
    key: "visitOpt_interval",
    value: function visitOpt_interval(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#interval_second.
  }, {
    key: "visitInterval_second",
    value: function visitInterval_second(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_escape.
  }, {
    key: "visitOpt_escape",
    value: function visitOpt_escape(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr.
  }, {
    key: "visitA_expr",
    value: function visitA_expr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_qual.
  }, {
    key: "visitA_expr_qual",
    value: function visitA_expr_qual(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_lessless.
  }, {
    key: "visitA_expr_lessless",
    value: function visitA_expr_lessless(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_or.
  }, {
    key: "visitA_expr_or",
    value: function visitA_expr_or(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_and.
  }, {
    key: "visitA_expr_and",
    value: function visitA_expr_and(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_between.
  }, {
    key: "visitA_expr_between",
    value: function visitA_expr_between(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_in.
  }, {
    key: "visitA_expr_in",
    value: function visitA_expr_in(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_unary_not.
  }, {
    key: "visitA_expr_unary_not",
    value: function visitA_expr_unary_not(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_isnull.
  }, {
    key: "visitA_expr_isnull",
    value: function visitA_expr_isnull(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_is_not.
  }, {
    key: "visitA_expr_is_not",
    value: function visitA_expr_is_not(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_compare.
  }, {
    key: "visitA_expr_compare",
    value: function visitA_expr_compare(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_like.
  }, {
    key: "visitA_expr_like",
    value: function visitA_expr_like(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_qual_op.
  }, {
    key: "visitA_expr_qual_op",
    value: function visitA_expr_qual_op(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_unary_qualop.
  }, {
    key: "visitA_expr_unary_qualop",
    value: function visitA_expr_unary_qualop(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_add.
  }, {
    key: "visitA_expr_add",
    value: function visitA_expr_add(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_mul.
  }, {
    key: "visitA_expr_mul",
    value: function visitA_expr_mul(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_caret.
  }, {
    key: "visitA_expr_caret",
    value: function visitA_expr_caret(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_unary_sign.
  }, {
    key: "visitA_expr_unary_sign",
    value: function visitA_expr_unary_sign(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_at_time_zone.
  }, {
    key: "visitA_expr_at_time_zone",
    value: function visitA_expr_at_time_zone(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_collate.
  }, {
    key: "visitA_expr_collate",
    value: function visitA_expr_collate(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#a_expr_typecast.
  }, {
    key: "visitA_expr_typecast",
    value: function visitA_expr_typecast(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#b_expr.
  }, {
    key: "visitB_expr",
    value: function visitB_expr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#c_expr_exists.
  }, {
    key: "visitC_expr_exists",
    value: function visitC_expr_exists(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#c_expr_expr.
  }, {
    key: "visitC_expr_expr",
    value: function visitC_expr_expr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#c_expr_case.
  }, {
    key: "visitC_expr_case",
    value: function visitC_expr_case(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#plsqlvariablename.
  }, {
    key: "visitPlsqlvariablename",
    value: function visitPlsqlvariablename(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_application.
  }, {
    key: "visitFunc_application",
    value: function visitFunc_application(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_expr.
  }, {
    key: "visitFunc_expr",
    value: function visitFunc_expr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_expr_windowless.
  }, {
    key: "visitFunc_expr_windowless",
    value: function visitFunc_expr_windowless(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_expr_common_subexpr.
  }, {
    key: "visitFunc_expr_common_subexpr",
    value: function visitFunc_expr_common_subexpr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xml_root_version.
  }, {
    key: "visitXml_root_version",
    value: function visitXml_root_version(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_xml_root_standalone.
  }, {
    key: "visitOpt_xml_root_standalone",
    value: function visitOpt_xml_root_standalone(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xml_attributes.
  }, {
    key: "visitXml_attributes",
    value: function visitXml_attributes(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xml_attribute_list.
  }, {
    key: "visitXml_attribute_list",
    value: function visitXml_attribute_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xml_attribute_el.
  }, {
    key: "visitXml_attribute_el",
    value: function visitXml_attribute_el(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#document_or_content.
  }, {
    key: "visitDocument_or_content",
    value: function visitDocument_or_content(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xml_whitespace_option.
  }, {
    key: "visitXml_whitespace_option",
    value: function visitXml_whitespace_option(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xmlexists_argument.
  }, {
    key: "visitXmlexists_argument",
    value: function visitXmlexists_argument(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xml_passing_mech.
  }, {
    key: "visitXml_passing_mech",
    value: function visitXml_passing_mech(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#within_group_clause.
  }, {
    key: "visitWithin_group_clause",
    value: function visitWithin_group_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#filter_clause.
  }, {
    key: "visitFilter_clause",
    value: function visitFilter_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#window_clause.
  }, {
    key: "visitWindow_clause",
    value: function visitWindow_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#window_definition_list.
  }, {
    key: "visitWindow_definition_list",
    value: function visitWindow_definition_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#window_definition.
  }, {
    key: "visitWindow_definition",
    value: function visitWindow_definition(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#over_clause.
  }, {
    key: "visitOver_clause",
    value: function visitOver_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#window_specification.
  }, {
    key: "visitWindow_specification",
    value: function visitWindow_specification(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_existing_window_name.
  }, {
    key: "visitOpt_existing_window_name",
    value: function visitOpt_existing_window_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_partition_clause.
  }, {
    key: "visitOpt_partition_clause",
    value: function visitOpt_partition_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_frame_clause.
  }, {
    key: "visitOpt_frame_clause",
    value: function visitOpt_frame_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#frame_extent.
  }, {
    key: "visitFrame_extent",
    value: function visitFrame_extent(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#frame_bound.
  }, {
    key: "visitFrame_bound",
    value: function visitFrame_bound(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_window_exclusion_clause.
  }, {
    key: "visitOpt_window_exclusion_clause",
    value: function visitOpt_window_exclusion_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#row.
  }, {
    key: "visitRow",
    value: function visitRow(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#explicit_row.
  }, {
    key: "visitExplicit_row",
    value: function visitExplicit_row(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#implicit_row.
  }, {
    key: "visitImplicit_row",
    value: function visitImplicit_row(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#sub_type.
  }, {
    key: "visitSub_type",
    value: function visitSub_type(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#all_op.
  }, {
    key: "visitAll_op",
    value: function visitAll_op(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#mathop.
  }, {
    key: "visitMathop",
    value: function visitMathop(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#qual_op.
  }, {
    key: "visitQual_op",
    value: function visitQual_op(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#qual_all_op.
  }, {
    key: "visitQual_all_op",
    value: function visitQual_all_op(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#subquery_Op.
  }, {
    key: "visitSubquery_Op",
    value: function visitSubquery_Op(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#expr_list.
  }, {
    key: "visitExpr_list",
    value: function visitExpr_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_arg_list.
  }, {
    key: "visitFunc_arg_list",
    value: function visitFunc_arg_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_arg_expr.
  }, {
    key: "visitFunc_arg_expr",
    value: function visitFunc_arg_expr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#type_list.
  }, {
    key: "visitType_list",
    value: function visitType_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#array_expr.
  }, {
    key: "visitArray_expr",
    value: function visitArray_expr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#array_expr_list.
  }, {
    key: "visitArray_expr_list",
    value: function visitArray_expr_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#extract_list.
  }, {
    key: "visitExtract_list",
    value: function visitExtract_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#extract_arg.
  }, {
    key: "visitExtract_arg",
    value: function visitExtract_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#unicode_normal_form.
  }, {
    key: "visitUnicode_normal_form",
    value: function visitUnicode_normal_form(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#overlay_list.
  }, {
    key: "visitOverlay_list",
    value: function visitOverlay_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#position_list.
  }, {
    key: "visitPosition_list",
    value: function visitPosition_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#substr_list.
  }, {
    key: "visitSubstr_list",
    value: function visitSubstr_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#trim_list.
  }, {
    key: "visitTrim_list",
    value: function visitTrim_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#in_expr_select.
  }, {
    key: "visitIn_expr_select",
    value: function visitIn_expr_select(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#in_expr_list.
  }, {
    key: "visitIn_expr_list",
    value: function visitIn_expr_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#case_expr.
  }, {
    key: "visitCase_expr",
    value: function visitCase_expr(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#when_clause_list.
  }, {
    key: "visitWhen_clause_list",
    value: function visitWhen_clause_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#when_clause.
  }, {
    key: "visitWhen_clause",
    value: function visitWhen_clause(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#case_default.
  }, {
    key: "visitCase_default",
    value: function visitCase_default(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#case_arg.
  }, {
    key: "visitCase_arg",
    value: function visitCase_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#columnref.
  }, {
    key: "visitColumnref",
    value: function visitColumnref(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#indirection_el.
  }, {
    key: "visitIndirection_el",
    value: function visitIndirection_el(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_slice_bound.
  }, {
    key: "visitOpt_slice_bound",
    value: function visitOpt_slice_bound(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#indirection.
  }, {
    key: "visitIndirection",
    value: function visitIndirection(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_indirection.
  }, {
    key: "visitOpt_indirection",
    value: function visitOpt_indirection(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_target_list.
  }, {
    key: "visitOpt_target_list",
    value: function visitOpt_target_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#target_list.
  }, {
    key: "visitTarget_list",
    value: function visitTarget_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#target_label.
  }, {
    key: "visitTarget_label",
    value: function visitTarget_label(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#target_star.
  }, {
    key: "visitTarget_star",
    value: function visitTarget_star(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#qualified_name_list.
  }, {
    key: "visitQualified_name_list",
    value: function visitQualified_name_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#qualified_name.
  }, {
    key: "visitQualified_name",
    value: function visitQualified_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#name_list.
  }, {
    key: "visitName_list",
    value: function visitName_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#name.
  }, {
    key: "visitName",
    value: function visitName(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#attr_name.
  }, {
    key: "visitAttr_name",
    value: function visitAttr_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#file_name.
  }, {
    key: "visitFile_name",
    value: function visitFile_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#func_name.
  }, {
    key: "visitFunc_name",
    value: function visitFunc_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#aexprconst.
  }, {
    key: "visitAexprconst",
    value: function visitAexprconst(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#xconst.
  }, {
    key: "visitXconst",
    value: function visitXconst(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#bconst.
  }, {
    key: "visitBconst",
    value: function visitBconst(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#fconst.
  }, {
    key: "visitFconst",
    value: function visitFconst(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#iconst.
  }, {
    key: "visitIconst",
    value: function visitIconst(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#sconst.
  }, {
    key: "visitSconst",
    value: function visitSconst(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#anysconst.
  }, {
    key: "visitAnysconst",
    value: function visitAnysconst(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_uescape.
  }, {
    key: "visitOpt_uescape",
    value: function visitOpt_uescape(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#signediconst.
  }, {
    key: "visitSignediconst",
    value: function visitSignediconst(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#roleid.
  }, {
    key: "visitRoleid",
    value: function visitRoleid(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#rolespec.
  }, {
    key: "visitRolespec",
    value: function visitRolespec(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#role_list.
  }, {
    key: "visitRole_list",
    value: function visitRole_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#colid.
  }, {
    key: "visitColid",
    value: function visitColid(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#table_alias.
  }, {
    key: "visitTable_alias",
    value: function visitTable_alias(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#type_function_name.
  }, {
    key: "visitType_function_name",
    value: function visitType_function_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#nonreservedword.
  }, {
    key: "visitNonreservedword",
    value: function visitNonreservedword(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#collabel.
  }, {
    key: "visitCollabel",
    value: function visitCollabel(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#identifier.
  }, {
    key: "visitIdentifier",
    value: function visitIdentifier(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#plsqlidentifier.
  }, {
    key: "visitPlsqlidentifier",
    value: function visitPlsqlidentifier(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#unreserved_keyword.
  }, {
    key: "visitUnreserved_keyword",
    value: function visitUnreserved_keyword(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#col_name_keyword.
  }, {
    key: "visitCol_name_keyword",
    value: function visitCol_name_keyword(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#type_func_name_keyword.
  }, {
    key: "visitType_func_name_keyword",
    value: function visitType_func_name_keyword(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#reserved_keyword.
  }, {
    key: "visitReserved_keyword",
    value: function visitReserved_keyword(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#builtin_function_name.
  }, {
    key: "visitBuiltin_function_name",
    value: function visitBuiltin_function_name(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#pl_function.
  }, {
    key: "visitPl_function",
    value: function visitPl_function(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#comp_options.
  }, {
    key: "visitComp_options",
    value: function visitComp_options(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#comp_option.
  }, {
    key: "visitComp_option",
    value: function visitComp_option(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#sharp.
  }, {
    key: "visitSharp",
    value: function visitSharp(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#option_value.
  }, {
    key: "visitOption_value",
    value: function visitOption_value(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_semi.
  }, {
    key: "visitOpt_semi",
    value: function visitOpt_semi(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#pl_block.
  }, {
    key: "visitPl_block",
    value: function visitPl_block(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_sect.
  }, {
    key: "visitDecl_sect",
    value: function visitDecl_sect(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_start.
  }, {
    key: "visitDecl_start",
    value: function visitDecl_start(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_stmts.
  }, {
    key: "visitDecl_stmts",
    value: function visitDecl_stmts(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#label_decl.
  }, {
    key: "visitLabel_decl",
    value: function visitLabel_decl(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_stmt.
  }, {
    key: "visitDecl_stmt",
    value: function visitDecl_stmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_statement.
  }, {
    key: "visitDecl_statement",
    value: function visitDecl_statement(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_scrollable.
  }, {
    key: "visitOpt_scrollable",
    value: function visitOpt_scrollable(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_cursor_query.
  }, {
    key: "visitDecl_cursor_query",
    value: function visitDecl_cursor_query(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_cursor_args.
  }, {
    key: "visitDecl_cursor_args",
    value: function visitDecl_cursor_args(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_cursor_arglist.
  }, {
    key: "visitDecl_cursor_arglist",
    value: function visitDecl_cursor_arglist(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_cursor_arg.
  }, {
    key: "visitDecl_cursor_arg",
    value: function visitDecl_cursor_arg(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_is_for.
  }, {
    key: "visitDecl_is_for",
    value: function visitDecl_is_for(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_aliasitem.
  }, {
    key: "visitDecl_aliasitem",
    value: function visitDecl_aliasitem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_varname.
  }, {
    key: "visitDecl_varname",
    value: function visitDecl_varname(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_const.
  }, {
    key: "visitDecl_const",
    value: function visitDecl_const(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_datatype.
  }, {
    key: "visitDecl_datatype",
    value: function visitDecl_datatype(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_collate.
  }, {
    key: "visitDecl_collate",
    value: function visitDecl_collate(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_notnull.
  }, {
    key: "visitDecl_notnull",
    value: function visitDecl_notnull(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_defval.
  }, {
    key: "visitDecl_defval",
    value: function visitDecl_defval(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#decl_defkey.
  }, {
    key: "visitDecl_defkey",
    value: function visitDecl_defkey(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#assign_operator.
  }, {
    key: "visitAssign_operator",
    value: function visitAssign_operator(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#proc_sect.
  }, {
    key: "visitProc_sect",
    value: function visitProc_sect(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#proc_stmt.
  }, {
    key: "visitProc_stmt",
    value: function visitProc_stmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_perform.
  }, {
    key: "visitStmt_perform",
    value: function visitStmt_perform(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_call.
  }, {
    key: "visitStmt_call",
    value: function visitStmt_call(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_expr_list.
  }, {
    key: "visitOpt_expr_list",
    value: function visitOpt_expr_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_assign.
  }, {
    key: "visitStmt_assign",
    value: function visitStmt_assign(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_getdiag.
  }, {
    key: "visitStmt_getdiag",
    value: function visitStmt_getdiag(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#getdiag_area_opt.
  }, {
    key: "visitGetdiag_area_opt",
    value: function visitGetdiag_area_opt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#getdiag_list.
  }, {
    key: "visitGetdiag_list",
    value: function visitGetdiag_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#getdiag_list_item.
  }, {
    key: "visitGetdiag_list_item",
    value: function visitGetdiag_list_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#getdiag_item.
  }, {
    key: "visitGetdiag_item",
    value: function visitGetdiag_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#getdiag_target.
  }, {
    key: "visitGetdiag_target",
    value: function visitGetdiag_target(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#assign_var.
  }, {
    key: "visitAssign_var",
    value: function visitAssign_var(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_if.
  }, {
    key: "visitStmt_if",
    value: function visitStmt_if(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_elsifs.
  }, {
    key: "visitStmt_elsifs",
    value: function visitStmt_elsifs(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_else.
  }, {
    key: "visitStmt_else",
    value: function visitStmt_else(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_case.
  }, {
    key: "visitStmt_case",
    value: function visitStmt_case(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_expr_until_when.
  }, {
    key: "visitOpt_expr_until_when",
    value: function visitOpt_expr_until_when(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#case_when_list.
  }, {
    key: "visitCase_when_list",
    value: function visitCase_when_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#case_when.
  }, {
    key: "visitCase_when",
    value: function visitCase_when(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_case_else.
  }, {
    key: "visitOpt_case_else",
    value: function visitOpt_case_else(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_loop.
  }, {
    key: "visitStmt_loop",
    value: function visitStmt_loop(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_while.
  }, {
    key: "visitStmt_while",
    value: function visitStmt_while(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_for.
  }, {
    key: "visitStmt_for",
    value: function visitStmt_for(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#for_control.
  }, {
    key: "visitFor_control",
    value: function visitFor_control(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_for_using_expression.
  }, {
    key: "visitOpt_for_using_expression",
    value: function visitOpt_for_using_expression(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_cursor_parameters.
  }, {
    key: "visitOpt_cursor_parameters",
    value: function visitOpt_cursor_parameters(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_reverse.
  }, {
    key: "visitOpt_reverse",
    value: function visitOpt_reverse(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_by_expression.
  }, {
    key: "visitOpt_by_expression",
    value: function visitOpt_by_expression(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#for_variable.
  }, {
    key: "visitFor_variable",
    value: function visitFor_variable(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_foreach_a.
  }, {
    key: "visitStmt_foreach_a",
    value: function visitStmt_foreach_a(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#foreach_slice.
  }, {
    key: "visitForeach_slice",
    value: function visitForeach_slice(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_exit.
  }, {
    key: "visitStmt_exit",
    value: function visitStmt_exit(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#exit_type.
  }, {
    key: "visitExit_type",
    value: function visitExit_type(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_return.
  }, {
    key: "visitStmt_return",
    value: function visitStmt_return(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_return_result.
  }, {
    key: "visitOpt_return_result",
    value: function visitOpt_return_result(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_raise.
  }, {
    key: "visitStmt_raise",
    value: function visitStmt_raise(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_stmt_raise_level.
  }, {
    key: "visitOpt_stmt_raise_level",
    value: function visitOpt_stmt_raise_level(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_raise_list.
  }, {
    key: "visitOpt_raise_list",
    value: function visitOpt_raise_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_raise_using.
  }, {
    key: "visitOpt_raise_using",
    value: function visitOpt_raise_using(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_raise_using_elem.
  }, {
    key: "visitOpt_raise_using_elem",
    value: function visitOpt_raise_using_elem(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_raise_using_elem_list.
  }, {
    key: "visitOpt_raise_using_elem_list",
    value: function visitOpt_raise_using_elem_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_assert.
  }, {
    key: "visitStmt_assert",
    value: function visitStmt_assert(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_stmt_assert_message.
  }, {
    key: "visitOpt_stmt_assert_message",
    value: function visitOpt_stmt_assert_message(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#loop_body.
  }, {
    key: "visitLoop_body",
    value: function visitLoop_body(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_execsql.
  }, {
    key: "visitStmt_execsql",
    value: function visitStmt_execsql(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_dynexecute.
  }, {
    key: "visitStmt_dynexecute",
    value: function visitStmt_dynexecute(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_execute_using.
  }, {
    key: "visitOpt_execute_using",
    value: function visitOpt_execute_using(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_execute_using_list.
  }, {
    key: "visitOpt_execute_using_list",
    value: function visitOpt_execute_using_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_execute_into.
  }, {
    key: "visitOpt_execute_into",
    value: function visitOpt_execute_into(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_open.
  }, {
    key: "visitStmt_open",
    value: function visitStmt_open(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_open_bound_list_item.
  }, {
    key: "visitOpt_open_bound_list_item",
    value: function visitOpt_open_bound_list_item(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_open_bound_list.
  }, {
    key: "visitOpt_open_bound_list",
    value: function visitOpt_open_bound_list(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_open_using.
  }, {
    key: "visitOpt_open_using",
    value: function visitOpt_open_using(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_scroll_option.
  }, {
    key: "visitOpt_scroll_option",
    value: function visitOpt_scroll_option(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_scroll_option_no.
  }, {
    key: "visitOpt_scroll_option_no",
    value: function visitOpt_scroll_option_no(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_fetch.
  }, {
    key: "visitStmt_fetch",
    value: function visitStmt_fetch(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#into_target.
  }, {
    key: "visitInto_target",
    value: function visitInto_target(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_cursor_from.
  }, {
    key: "visitOpt_cursor_from",
    value: function visitOpt_cursor_from(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_fetch_direction.
  }, {
    key: "visitOpt_fetch_direction",
    value: function visitOpt_fetch_direction(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_move.
  }, {
    key: "visitStmt_move",
    value: function visitStmt_move(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_close.
  }, {
    key: "visitStmt_close",
    value: function visitStmt_close(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_null.
  }, {
    key: "visitStmt_null",
    value: function visitStmt_null(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_commit.
  }, {
    key: "visitStmt_commit",
    value: function visitStmt_commit(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_rollback.
  }, {
    key: "visitStmt_rollback",
    value: function visitStmt_rollback(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#plsql_opt_transaction_chain.
  }, {
    key: "visitPlsql_opt_transaction_chain",
    value: function visitPlsql_opt_transaction_chain(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#stmt_set.
  }, {
    key: "visitStmt_set",
    value: function visitStmt_set(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#cursor_variable.
  }, {
    key: "visitCursor_variable",
    value: function visitCursor_variable(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#exception_sect.
  }, {
    key: "visitException_sect",
    value: function visitException_sect(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#proc_exceptions.
  }, {
    key: "visitProc_exceptions",
    value: function visitProc_exceptions(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#proc_exception.
  }, {
    key: "visitProc_exception",
    value: function visitProc_exception(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#proc_conditions.
  }, {
    key: "visitProc_conditions",
    value: function visitProc_conditions(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#proc_condition.
  }, {
    key: "visitProc_condition",
    value: function visitProc_condition(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_block_label.
  }, {
    key: "visitOpt_block_label",
    value: function visitOpt_block_label(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_loop_label.
  }, {
    key: "visitOpt_loop_label",
    value: function visitOpt_loop_label(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_label.
  }, {
    key: "visitOpt_label",
    value: function visitOpt_label(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_exitcond.
  }, {
    key: "visitOpt_exitcond",
    value: function visitOpt_exitcond(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#any_identifier.
  }, {
    key: "visitAny_identifier",
    value: function visitAny_identifier(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#plsql_unreserved_keyword.
  }, {
    key: "visitPlsql_unreserved_keyword",
    value: function visitPlsql_unreserved_keyword(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#sql_expression.
  }, {
    key: "visitSql_expression",
    value: function visitSql_expression(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#expr_until_then.
  }, {
    key: "visitExpr_until_then",
    value: function visitExpr_until_then(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#expr_until_semi.
  }, {
    key: "visitExpr_until_semi",
    value: function visitExpr_until_semi(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#expr_until_rightbracket.
  }, {
    key: "visitExpr_until_rightbracket",
    value: function visitExpr_until_rightbracket(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#expr_until_loop.
  }, {
    key: "visitExpr_until_loop",
    value: function visitExpr_until_loop(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#make_execsql_stmt.
  }, {
    key: "visitMake_execsql_stmt",
    value: function visitMake_execsql_stmt(ctx) {
      return this.visitChildren(ctx);
    }

    // Visit a parse tree produced by PostgreSQLParser#opt_returning_clause_into.
  }, {
    key: "visitOpt_returning_clause_into",
    value: function visitOpt_returning_clause_into(ctx) {
      return this.visitChildren(ctx);
    }
  }]);
  return PostgreSQLParserVisitor;
}(_antlr["default"].tree.ParseTreeVisitor);
exports["default"] = PostgreSQLParserVisitor;
